File: notifications-api-typescript/src/processor.ts
import { Notification, NotificationStatus } from "./models.js";
import * as storage from "./storage.js";
import { send as sendEmail } from "./providers/emailProvider.js";
import { send as sendSms } from "./providers/smsProvider.js";
import { send as sendPush } from "./providers/pushProvider.js";

export class NotificationProcessor {
  sendOne(n: Notification): void {
    n.status = NotificationStatus.PROCESSING;
    n.attempts++;
    n.lastAttemptAt = new Date();
    if (n.targetChannels.length === 0) {
      n.status = NotificationStatus.FAILED;
      n.lastError = "No target channels";
      return;
    }

    for (const target of n.targetChannels) {
      let response;

      if (target.type === "email") {
        response = sendEmail({
          recipient: target.value,
          message: n.message,
        });
      } else if (target.type === "sms") {
        response = sendSms({
          recipient: target.value,
          message: n.message,
        });
      } else if (target.type === "push") {
        response = sendPush({
          recipient: target.value,
          message: n.message,
        });
      } else {
        n.status = NotificationStatus.FAILED;
        n.lastError = "Unknown channel";
        return;
      }

      n.lastError = response.Message;
    }

    n.status = NotificationStatus.SENT;
  }

  sendAll(): void {
    const pending = storage.getAll().filter((n) => n.status === NotificationStatus.PENDING || n.status === NotificationStatus.RETRY_PENDING);
    for (const n of pending) {
      this.sendOne(n);
    }
  }
}

File: notifications-api-python/src/processor.py
from datetime import datetime

import storage
from models import PENDING, PROCESSING, SENT, FAILED
from providers.email_provider import send as send_email
from providers.sms_provider import send as send_sms
from providers.push_provider import send as send_push


class NotificationProcessor:
    def send_one(self, n):
        n.status = PROCESSING
        n.attempts += 1
        n.lastAttemptAt = datetime.now().isoformat()

        if not n.targetChannels:
            n.status = FAILED
            n.lastError = "No target channels"
            return
        target = n.targetChannels[0]

        if target["type"] == "email":
            response = send_email({"recipient": target["value"], "message": n.message})
        elif target["type"] == "sms":
            response = send_sms({"recipient": target["value"], "message": n.message})
        elif target["type"] == "push":
            response = send_push({"recipient": target["value"], "message": n.message})
        else:
            n.status = FAILED
            n.lastError = "Unknown channel"
            return

        n.status = SENT
        n.lastError = response["Message"]

    def send_all(self):
        pending = [n for n in storage.get_all() if n.status == PENDING]
        for n in pending:
            self.send_one(n)
